docs/docs/reference/extend/extension-methods.md

```diff
@@ -0,0 +1,254 @@
+
+### Translation of Extension Methods
+
+Extension methods are methods that have a parameter clause in front of the defined
+identifier. They translate to methods where the leading parameter section is moved
+to after the defined identifier. So, the definition of `circumference` above translates
+to the plain method, and can also be invoked as such:
+```scala
+def circumference(c: Circle): Double = c.radius * math.Pi * 2
+
+assert(circle.circumference == CircleOps.circumference(circle))
+```
+
+### Translation of Calls to Extension Methods
+
+The rules for resolving a selection `e.m` are augmented as follows: If `m` is not a
+member of the type `T` of `e`, and there is an implicit value `i`
+in either the current scope or in the implicit scope of `T`, and `i` defines an extension
+method named `m`, then `e.m` is expanded to `i.m(e)`. This expansion is attempted at the time where the compiler also tries an implicit conversion from `T` to a type containing `m`. If there is more than one way
+of expanding, an ambiguity error results.
+
+So `circle.circumference` translates to `CircleOps.circumference(circle)`, provided
+`circle` has type `Circle` and `CircleOps` is an eligible implicit (i.e. it is visible at the point of call or it is defined in the companion object of `Circle`).
+
+### Extended Types
+
+Extension methods can be added to arbitrary types. For instance, the following
+object adds a `longestStrings` extension method to a `Seq[String]`:
+
+```scala
+implicit object StringOps {
+  def (xs: Seq[String]).longestStrings = {
+    val maxLength = xs.map(_.length).max
+    xs.filter(_.length == maxLength)
+  }
+}
+```
+
+### Generic Extensions
+
+The previous example extended a specific instance of a generic type. It is also possible
+to extend a generic type by adding type parameters to an extension method:
+
+```scala
+implicit object ListOps {
+  def (xs: List[T]).second[T] = xs.tail.head
+}
+```
+
+or:
+
+
+```scala
+implicit object ListListOps {
+  def (xs: List[List[T]]).flattened[T] = xs.foldLeft[List[T]](Nil)(_ ++ _)
+}
+```
+
+As usual, type parameters of the extension method follow the defined method name. Nevertheless, such type parameters can already be used in the parameter clause that precedes
+the defined method name.
+
+### A Larger Example
+
+As a larger example, here is a way to define constructs for checking arbitrary postconditions using `ensuring` so that the checked result can be referred to simply by `result`. The example combines opaque aliases, implicit function types, and extensions to provide a zero-overhead abstraction.
+
+```scala
+object PostConditions {
+  opaque type WrappedResult[T] = T
+
+  private object WrappedResult {
+    def wrap[T](x: T): WrappedResult[T] = x
+    def unwrap[T](x: WrappedResult[T]): T = x
+  }
+
+  def result[T](implicit er: WrappedResult[T]): T = WrappedResult.unwrap(er)
+
+  implicit object Ensuring {
+    def (x: T).ensuring(condition: implicit WrappedResult[T] => Boolean): T = {
```

Comment:
This is the only bit of this proposal that I'm not getting.  How, in this declaration, does the compiler know that `T` (in `def (x: T).ensuring(...)`) is generic?  I'm not seeing any obvious distinction between `T` as concrete type and `T` as type parameter here...

Reply:
Oops, no wonder you were not getting that! There's a type parameter missing. That's the problem of posting example code without having a compiler to check it...